### PR TITLE
Step-7: Filtering Tasks

### PR DESCRIPTION
**Pull request by fza-wtag**
## Description
[Step-7](https://trello.com/c/BhhLxwTY)
This PR introduces all changes regarding implementing filtering the tasks.

## Checklist
- [x]  By default, it shows all the tasks. 
- [x]  Upon clicking on Incomplete, it shows only the incomplete tasks and the same for Complete and All.
- [x]  Filtering also works well with pagination.

## Attachments
https://user-images.githubusercontent.com/119289799/217800061-6f313e2b-07e4-4630-8b6d-a0488b9d1b93.mov

### CODE REVIEW

File: todos/src/components/FilterButtons.js [hunk 1]
```diff
@@ -1,13 +1,56 @@
-import React from 'react'
+import React from "react";
+import { setFilter } from "actions/index";
+import { useDispatch, useSelector } from "react-redux";
+
+const FilterButtons = () => {
+  const dispatch = useDispatch();
+  const filter = useSelector((state) => state.filterReducer.filter);
+  const ListLength = useSelector((state) => state.todoReducers.list).length;
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:8]:
```suggestion
  const listLength = useSelector((state) => state.todoReducers.list).length;
```

File: todos/src/components/FilterButtons.js [hunk 2]
```diff
@@ -1,13 +1,56 @@
...
+
+  const handleAllTodos = () => {
+    dispatch(setFilter("all"));
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:11]: Can we use const here?
```suggestion
    dispatch(setFilter(ALL));
```

File: todos/src/components/FilterButtons.js [hunk 3]
```diff
@@ -1,13 +1,56 @@
...
+  };
+  const handleIncompleteTodos = () => {
+    dispatch(setFilter("incomplete"));
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:14]: ditto

File: todos/src/components/FilterButtons.js [hunk 4]
```diff
@@ -1,13 +1,56 @@
...
+  };
+  const handleCompleteTodos = () => {
+    dispatch(setFilter("complete"));
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:17]: ditto

File: todos/src/components/Todos.js [hunk 5]
```diff
@@ -5,24 +5,31 @@ import "styles/todos.css";
...
   const currentPage = useSelector(
     (state) => state.currentPageReducer.currentPage
   );
-  const displayedTodoList = list.slice(0, PER_PAGE * currentPage);
+  const filteredTodos =
+    filter === "all"
+      ? list
+      : filter === "incomplete"
+      ? list.filter((todo) => !todo.isCompleted)
+      : list.filter((todo) => todo.isCompleted);
```
Comment (hhu-wtag) [inline todos/src/components/Todos.js:24]: This is hard to read. Let's do it this way
```suggestion
    if( filter === ALL ){
      return list
    }
    
    if( filter === INCOMPLETE ) {
      return list.filter((todo) => !todo.isCompleted)
    }

    return list.filter((todo) => todo.isCompleted)
```

File: todos/src/components/Todos.js [hunk 6]
```diff
@@ -32,14 +39,14 @@ function Todos() {
...
               onEdit={elem.onEdit}
             />
-          );
+          ) : null;
         })}
       </div>
       {list.length === 0 && !isAddTaskVisible && <EmptyTaskList />}
-      {currentPage * PER_PAGE < list.length ? (
-        <LoadMoreBtn type={"Load More"} />
+      {currentPage * PER_PAGE < filteredTodos.length ? (
```
Comment (hhu-wtag) [inline todos/src/components/Todos.js:46]: Can we port this to a computed variable to make the markup less complex?

**Review by hhu-wtag — CHANGES_REQUESTED** [review]
- [ ] Let's use constants

File: todos/src/components/FilterButtons.js (hunk 2 above)
Comment (fza-wtag) [reply todos/src/components/FilterButtons.js:11]: Yes, @hhu-wtag Bhai. Doing it right now.

File: todos/src/components/Todos.js (hunk 5 above)
Comment (fza-wtag) [reply todos/src/components/Todos.js:24]: Alright, @hhu-wtag Bhai. Done.

File: todos/src/components/Todos.js (hunk 6 above)
Comment (fza-wtag) [reply todos/src/components/Todos.js]: Yes, @hhu-wtag Bhai. We've already done that on our pagination branch, so I resolved the conflicts, and now all the markup in both branches is correct, with less complexity. Thank you, Bhai for pointing it out again.

**Review by hhu-wtag — APPROVED** [review]

File: todos/src/actions/index.js [hunk 7]
```diff
@@ -84,4 +84,11 @@ export const updateCurrentPage = (page) => {
...
     
   };
+};
+
+export const setFilter = (filter) => {
+  return {
+    type: "SET_FILTER",
+    filter,
+  };
 };
```
Comment (ray-wtag) [inline todos/src/actions/index.js:94]: prettier please

File: todos/src/components/FilterButtons.js [hunk 8]
```diff
@@ -1,13 +1,55 @@
...
 
-function FilterButtons() {
   return (
-    <div className="btn__progress_button">
-      <button>All</button>
-      <button>Incomplete</button>
-      <button>Complete</button>
+    <div
+      className={`btn__progress_button ${
+        listLength === 0 ? "btn__progress_button--disable" : null
```
Comment (ray-wtag) [inline todos/src/components/FilterButtons.js:24]: please use classnames npm package here

File: todos/src/components/FilterButtons.js [hunk 9]
```diff
@@ -1,13 +1,55 @@
...
+      }`}
+    >
+      <button
+        className={
+          filter === ALL && listLength !== 0
```
Comment (ray-wtag) [inline todos/src/components/FilterButtons.js:29]: ditto

File: todos/src/components/FilterButtons.js [hunk 10]
```diff
@@ -1,13 +1,55 @@
...
+            ? "btn__progress_button--disable"
+            : null
+        }
+        onClick={handleAllTodos}
+      >
+        All
+      </button>
+      <button
+        className={
+          filter === INCOMPLETE ? "btn__progress_button--disable" : null
```
Comment (ray-wtag) [inline todos/src/components/FilterButtons.js:39]: ditto

File: todos/src/components/FilterButtons.js [hunk 11]
```diff
@@ -1,13 +1,55 @@
...
+        }
+        onClick={handleIncompleteTodos}
+      >
+        Incomplete
+      </button>
+      <button
+        className={filter === COMPLETE ? "btn__progress_button--disable" : null}
```
Comment (ray-wtag) [inline todos/src/components/FilterButtons.js:46]: ditto

File: todos/src/components/FilterButtons.js (hunk 8 above)
Comment (fza-wtag) [reply todos/src/components/FilterButtons.js:24]: @ray-wtag Bhai, resolved [here](https://github.com/fza-wtag/todo-app-react/pull/9#discussion_r1111192324) in another PR.

**Review by ray-wtag — APPROVED** [review]

**Pull request merged**